executor/executor_test.go

```diff
@@ -194,6 +225,749 @@ func TestStop(t *testing.T) {
+		numKeys      = 5
+		numTxs       = 100_000
+		conflictKeys = make([]string, 0, numKeys)
+		blocking     = set.Set[int]{}
+		l            sync.Mutex
+		completed    = make([]int, 0, numTxs)
+		answers      = make([][]int, 5)
+		e            = New(numTxs, 4, nil)
+		slow         = make(chan struct{})
+	)
+
+	// make [answers] matrix
+	answers[0] = generateNumbers(0)
+	answers[1] = generateNumbers(20000)
+	answers[2] = generateNumbers(40000)
+	answers[3] = generateNumbers(60000)
+	answers[4] = generateNumbers(80000)
+
+	// randomly wait on certain txs
+	for {
+		if blocking.Len() == 10000 {
+			break
+		}
+		blocking.Add(rand.Intn(100_000)) //nolint:gosec
+	}
+
+	// make the conflict keys
+	for k := 0; k < numKeys; k++ {
+		conflictKeys = append(conflictKeys, ids.GenerateTestID().String())
+	}
+
+	// Generate txs
+	for i := 0; i < numTxs; i++ {
+		s := make(state.Keys, (numKeys + 1))
+		for j := 0; j < numKeys; j++ {
+			mode := (i / 10000) % 2
+			switch mode {
+			case 0:
+				s.Add(conflictKeys[j], state.Write)
+			case 1:
+				s.Add(conflictKeys[j], state.Read)
+			}
+		}
+
+		// pass into executor
+		ti := i
+		e.Run(s, func() error {
+			shouldWait := false
+			for num := range blocking {
+				if num == ti {
+					shouldWait = true
+					break
+				}
+			}
+			if shouldWait {
+				<-slow
+			}
+
+			l.Lock()
+			completed = append(completed, ti)
+			l.Unlock()
+			return nil
+		})
+	}
+	for i := 0; i < 1000; i++ {
+		slow <- struct{}{}
+	}
+	close(slow)
+	require.NoError(e.Wait())
+	require.Len(completed, numTxs)
+	require.Equal(answers[0], completed[0:9999])
+	require.Equal(answers[1], completed[20000:29999])
+	require.Equal(answers[2], completed[40000:49999])
+	require.Equal(answers[3], completed[60000:69999])
+	require.Equal(answers[4], completed[80000:89999])
+}
+
+// Random mix of unique and conflicting keys. We just
+// need to see if we don't halt in this test
+func TestLargeRandomReadsAndWrites(t *testing.T) {
+	rand.Seed(5)
+
+	var (
+		require      = require.New(t)
+		numKeys      = 10
+		numTxs       = 100_000
+		conflictKeys = make([]string, 0, numKeys)
+		blocking     = set.Set[int]{}
+		l            sync.Mutex
+		completed    = make([]int, 0, numTxs)
+		e            = New(numTxs, 4, nil)
+		slow         = make(chan struct{})
+	)
+
+	// randomly wait on certain txs
+	for {
+		if blocking.Len() == 10000 {
+			break
+		}
+		blocking.Add(rand.Intn(100_000)) //nolint:gosec
+	}
+
+	// make the conflict keys
+	for k := 0; k < numKeys; k++ {
+		conflictKeys = append(conflictKeys, ids.GenerateTestID().String())
+	}
+
+	// Generate txs
+	for i := 0; i < numTxs; i++ {
+		// mix of unique and conflict keys in each tx
+		s := make(state.Keys, (numKeys + 1))
+
+		// random size of conflict keys to add
+		min := 1
+		max := 6
+		setSize := rand.Intn(max-min+1) + min             //nolint:gosec
+		randomConflictingKeys := set.NewSet[int](setSize) // indices of [conflictKeys]
+		for {
+			if randomConflictingKeys.Len() == setSize {
```

Comment:
typically, you'd just make this the termination condition of the for loop...this `if` immediately inside of a `for` is odd

Reply:
I think you do this a few places